# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

<b>Answer: </b>

<b>1.</b>
This paper is about a new framework to tidy messy datasets. The goal of this research is to make data cleaning as easy and effective as possible. 

<b>2.</b>
The tidy data standard is intended to accomplish several things, such as facilitating the exploration and analysis of data, reduce data munging, and bridging the gap between the organization and analysis of data. By having the tidy data standard, analysts can keep datasets consistent, and collaboration can be streamlined between people. 


<b>3.</b>
"Like families, tidy datasets are all alike but every messy dataset is messy in its own way". This sentence is basically saying that just as families have similar characteristics that make them happy, datasets have things that make them "tidy" as well. In contrast, just as every unhappy family has problems, every messy dataset has some problems as well. 

"For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general". This sentence is basically saying that for sp0ecific datasets, like those for houses, it is easy to figure out the observations and the attributes/features that describe those observations. However, it is very difficult to find a universal definition for observations and features because different datasets from different domains have different kinds of observations and features. 

<b>4.</b>

Values:
Values are defined as something that belongs to a variable and an observation. More specifically, the values are either numbers if they are quantitative and strings if they are qualitative.

Variables:
These are the things that contain the values. They contain the values that define a trait (such as height, temperature, and duration). As a result, they can hold both quantitative and qualitative data. 

Observations:
Observations are also things that values belong to. The observation contains all values measured on the same unit across different attributes. 


<b>5.</b>

Tidy data is defined as "a standard way of mapping the meaning of a dataset to its structure" according to section 2.3. In tidy data, each variable forms a column, each observation forms a row, each type of observational unit forms a table.

<b>6.</b>
The 5 most common problems with messy datasets are as follows: Column headers are values, not variable names. Multiple variables are stored in one column. Variables are stored in both rows and columns. Multiple types of observational units are stored in the same table. A single observational unit is stored in multiple tables.

The data in table 4 is messy because the column headers are values and not variable names. "Melting" a dataset is to turn the columns of the dataset into rows. This can help clean the data because it can turn the column headers that are values into rows instead.


<b>7.</b>

The reason that table 11 is messy is because there are multiple types in one table. For example, the fact that there is a column for every day in the month makes it messy along with the multiple types (tmax and tmin) in one column. Table 12 is tidy and molten because it consolidates the dates and values into two columns (date and value) for the molten data. The tidy data separates the multiple types in one table and makes two columns from that (tmax and tmin), eliminating the need of an element/value column as well.

<b>8.</b>
The chicken and egg problem is the interdependence between tidy data and the tools that work with it. If tidy data is only as useful as the tools that work with it, then the tools will be too specialized to the structure of the tidy data. This leads to a situation where changing the data structures or tools won't improve the workflow. This results in a state where it seems like the current methodology is optimal but is actually suboptimal.



**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [2]:
##NOTE: PART 1 HERE
import pandas as pd

air_data = pd.read_csv('data/airbnb_hw.csv')
# air_data.describe()
# air_data.head()
# print(air_data['Price'].astype(str))
print('''
There seems to be a problem with the Price variable as it is being stored as a string. If it weren't a string, then the .describe would have gotten it.
''')

#changing all the prices to numeric
air_data['Price'] = pd.to_numeric(air_data['Price'].str.replace(',', ''), errors='coerce')
# print(air_data)
print(air_data.describe())

print("""
      The price variable has been cleaned to not include commas. 
      It has also been changed so that it is of a numeric type instead of a string.
      The results of these changes are beneficial as I can also use statistics on the Prices variable now as well.""")







There seems to be a problem with the Price variable as it is being stored as a string. If it weren't a string, then the .describe would have gotten it.

            Host Id  Review Scores Rating (bin)       Zipcode          Beds  \
count  3.047800e+04                22155.000000  30344.000000  30393.000000   
mean   1.273171e+07                   90.738659  10584.854831      1.530089   
std    1.190270e+07                    9.059519    921.299397      1.015359   
min    5.000000e+02                   20.000000   1003.000000      0.000000   
25%    2.701298e+06                   85.000000  10017.000000      1.000000   
50%    8.551693e+06                   90.000000  10065.000000      1.000000   
75%    2.120617e+07                  100.000000  11216.000000      2.000000   
max    4.303307e+07                  100.000000  99135.000000     16.000000   

       Number of Records  Number Of Reviews         Price  \
count            30478.0       30478.000000  30478.000000   
mean        

In [3]:
# NOTE: PART 2 HERE
sharks_data = pd.read_csv('data/sharks.csv')

type_counts = sharks_data["Type"].value_counts()

print(type_counts)
print("""
    From what I can see in the values for the type column, there seem to several related types that I plan on combining into one type. 
    For example, The Boat, Boating, and Boatomg types all seem to be related so I am going to convert them into the Boating type.
    The Invalid, Questionable, Unconfirmed, Unverified, and Under investigation columns also seem to be similar so I will combine them into the Uncertain type.
    The other types seem different enough so I will keep them as is. 
    The final result of the cleaned data can be seen below.
""")
sharks_data["Type"].replace(["Boat", "Boatomg"], "Boating", inplace=True)

uncertain_types = ["Invalid", "Questionable", "Unconfirmed", "Unverified", "Under investigation"]
sharks_data["Type"].replace(uncertain_types, "Uncertain", inplace=True)

print(sharks_data)
new_type_counts = sharks_data["Type"].value_counts()
print(new_type_counts)


Unprovoked             4716
Provoked                593
Invalid                 552
Sea Disaster            239
Watercraft              142
Boat                    109
Boating                  92
Questionable             10
Unconfirmed               1
Unverified                1
Under investigation       1
Boatomg                   1
Name: Type, dtype: int64

    From what I can see in the values for the type column, there seem to several related types that I plan on combining into one type. 
    For example, The Boat, Boating, and Boatomg types all seem to be related so I am going to convert them into the Boating type.
    The Invalid, Questionable, Unconfirmed, Unverified, and Under investigation columns also seem to be similar so I will combine them into the Uncertain type.
    The other types seem different enough so I will keep them as is. 
    The final result of the cleaned data can be seen below.

      index   Case Number                  Date    Year        Type  \
0         

C:\Users\Hieut\AppData\Local\Temp\ipykernel_37416\2641986835.py:2: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  sharks_data = pd.read_csv('data/sharks.csv')


In [4]:
# NOTE: PART 3 HERE
import numpy as np
pretrial_data = pd.read_csv("data/VirginiaPretrialData2017.csv")


print(pretrial_data["WhetherDefendantWasReleasedPretrial"].value_counts())

print("""As seen in the dataset values, there are a lot of 9s that do not really mean much in the dataset. 
      Below, you can see the change from a 9 to np.nan """)

# changing the 9s to nans
pretrial_data['WhetherDefendantWasReleasedPretrial'] = pretrial_data['WhetherDefendantWasReleasedPretrial'].replace(9, np.nan)

print(pretrial_data["WhetherDefendantWasReleasedPretrial"].value_counts())
print(pretrial_data["WhetherDefendantWasReleasedPretrial"].unique())




C:\Users\Hieut\AppData\Local\Temp\ipykernel_37416\4156441379.py:3: DtypeWarning: Columns (1,4,5,7,79,80,81,82,83,84,108,163,164,165,166,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,270,271,272,273,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,301,302,303,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,

1    19154
0     3801
9       31
Name: WhetherDefendantWasReleasedPretrial, dtype: int64
As seen in the dataset values, there are a lot of 9s that do not really mean much in the dataset. 
      Below, you can see the change from a 9 to np.nan 
1.0    19154
0.0     3801
Name: WhetherDefendantWasReleasedPretrial, dtype: int64
[nan  0.  1.]


In [8]:
# NOTE: PART 4


# 4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as 
# well as you can, and explain the choices you make.
# (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)
print("""The main choice I made here was concerning the 9s and the 4s 
      in the SentenceTypeAllChargesAtConvictionInContactEvent variable.
      What I noticed that the corresponding rows in the ImposedSentenceAllChargeInContactEvent
      variable did not have a value to them. As a result, the main change that I am implementing is changing
      all of those blank cells to nans. This should be an easy fix by just checking the 4s,9s in the 
      SentenceTypeAllChargesAtConvictionInContactEvent variable and changing the corresponding row in 
      ImposedSentenceAllChargeInContactEvent to nan""")

new_pretrial_data = pd.read_csv("data/VirginiaPretrialData2017.csv")
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html
new_pretrial_data.loc[new_pretrial_data['SentenceTypeAllChargesAtConvictionInContactEvent'].isin([4, 9]), 'ImposedSentenceAllChargeInContactEvent'] = np.nan
missing_vals = new_pretrial_data['ImposedSentenceAllChargeInContactEvent'].isnull().sum()

print(missing_vals)
print(new_pretrial_data)



The main choice I made here was concerning the 9s and the 4s 
      in the SentenceTypeAllChargesAtConvictionInContactEvent variable.
      What I noticed that the corresponding rows in the ImposedSentenceAllChargeInContactEvent
      variable did not have a value to them. As a result, the main change that I am implementing is changing
      all of those blank cells to nans. This should be an easy fix by just checking the 4s,9s in the 
      SentenceTypeAllChargesAtConvictionInContactEvent variable and changing the corresponding row in 
      ImposedSentenceAllChargeInContactEvent to nan


C:\Users\Hieut\AppData\Local\Temp\ipykernel_37416\2577742434.py:15: DtypeWarning: Columns (1,4,5,7,79,80,81,82,83,84,108,163,164,165,166,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,270,271,272,273,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,301,302,303,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390

9053
      InternalStudyID REQ_REC# Defendant_Sex Defendant_Race  \
0            ADI00001        1             M              W   
1            ADI00007        3             M              B   
2            ADI00008        4             M              W   
3            CDI00036        6             M              B   
4            CDI00038        7             F              W   
...               ...      ...           ...            ...   
22981        SJA00343    22839             M              B   
22982        SJA00344                      F              W   
22983        SJA00345    22841             M              B   
22984        SJA00346    22842             F              B   
22985        SJA00347    22843             M              B   

      Defendant_BirthYear Defendant_Age  Defendant_AgeGroup  \
0                    1986            31                   3   
1                    1956            60                   6   
2                    1990            27          

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.